### PR TITLE
[Search Pipelines] Add request-scoped state shared between processors

### PR DESCRIPTION
**Pull request by msfroh**
To handle cases where multiple search pipeline processors need to share information, we will allocate a Map<String, Object> for the lifetime of the request and pass it to each processor to get/set values.

### Description
To handle cases where multiple search pipeline processors need to share information, we will allocate a `Map<String, Object>` for the lifetime of the request and pass it to each processor to get/set values.

### Related Issues
https://github.com/opensearch-project/OpenSearch/issues/6722

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/helpers/SearchResponseUtil.java
Comment (austintlee) [inline modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/helpers/SearchResponseUtil.java:32]: Shouldn't we use `InternalSearchResponse` here?

https://github.com/opensearch-project/OpenSearch/blob/61c5f173b9b6dba7de84423be5a9db283e25291b/server/src/main/java/org/opensearch/action/search/SearchResponse.java#L96C1-L96C1

**Comment by austintlee** [comment]
@msfroh This looks really good.  Let me know if there's anything I can do to help get this over the line.  I can help with writing unit tests, e.g.

File: modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/helpers/SearchResponseUtil.java
Comment (msfroh) [reply modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/helpers/SearchResponseUtil.java:32]: Yes, that would be safer -- my understanding is that `InternalSearchResponse` implements `Writeable`, which is required for the whole `SearchResponse` to be `Writeable`.

While the YamlRestTests succeed with this implementation, I think it would run into problems if called from a transport client. It would probably break cross-cluster search.

I'll make that change in my next commit.

File: modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/TruncateHitsResponseProcessor.java
Comment (noCharger) [inline modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/TruncateHitsResponseProcessor.java:55]: This condition determines if the size should be fetched from the context. I would add a comment explaining why a negative targetSize indicates this behavior.

File: modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/TruncateHitsResponseProcessor.java
Comment (noCharger) [inline modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/TruncateHitsResponseProcessor.java:75]: Should we consider including this in the SearchResponseUtil as well?

File: modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/helpers/SearchResponseUtil.java
Comment (noCharger) [inline modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/helpers/SearchResponseUtil.java:50]: How about adding this method to the SearchResponse class as either a copy constructor or a rewritten function?

File: modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/OversampleRequestProcessor.java
Comment (msfroh) [reply modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/OversampleRequestProcessor.java:46]: No, never

File: server/src/main/java/org/opensearch/search/pipeline/StatefulSearchRequestProcessor.java
Comment (msfroh) [reply server/src/main/java/org/opensearch/search/pipeline/StatefulSearchRequestProcessor.java:18]: `SearchRequestProcessor` is not necessarily stateless, but I couldn't break the existing interface (that doesn't take state). One could just implement `SearchRequestProcessor` and implement the method that takes state, while throwing `UnsupportedOperationException` for the method that doesn't take state.

The sub-interfaces are just there to make it easier for developers. Extend the sub-interface and the compiler will force you to implement the method that takes state (since the default implementation from the parent has been cleared).

File: modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/helpers/SearchResponseUtil.java
Comment (austintlee) [reply modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/helpers/SearchResponseUtil.java:50]: We had a similar discussion on this here - https://github.com/opensearch-project/OpenSearch/pull/9379#discussion_r1305705244.

**Review by noCharger — APPROVED** [review]
This is fantastic! Approved, and please update the gradle check by adding java documentation.

> Task :modules:search-pipeline-common:missingJavadoc FAILED

File: server/src/main/java/org/opensearch/search/pipeline/Processor.java
Comment (msfroh) [reply server/src/main/java/org/opensearch/search/pipeline/Processor.java:30]: It should never have been pushed. 

It's a leftover from an earlier revision of https://github.com/opensearch-project/OpenSearch/pull/8164 that I forgot to remove after I changed to a different approach.

**Comment by macohen** [comment]
we're going to need to push this to a 2.11 release. Was anyone counting on anything in this particular PR for a 2.10 release or can you work with it once it gets merged into main and then we release for 2.11? 

cc: @austintlee

**Comment by austintlee** [comment]
2.11 is fine for me.  Thanks for checking with me.

**Comment by austintlee** [comment]
Just for my understanding, what work is remaining?  Is it that we didn't get enough reviewers on this?

**Comment by msfroh** [comment]
> Just for my understanding, what work is remaining? Is it that we didn't get enough reviewers on this?

Essentially yes, I think. Lots of changes going in to 2.10 and only so many maintainers with knowledge of this part of the code.

I'll try to do my part to review more stuff so that the other maintainers have more time to review my stuff going forward.

File: modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/OversampleRequestProcessor.java
Comment (navneet1v) [inline modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/OversampleRequestProcessor.java:47]: Question: as size is optional parameter. when control this code path does the size already set to default value of 10?

File: modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/OversampleRequestProcessor.java
Comment (navneet1v) [inline modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/OversampleRequestProcessor.java:49]: should we have some validation here so that size is not going above like 10k which is max size value for OpenSearch?

File: server/src/main/java/org/opensearch/search/pipeline/Pipeline.java
Comment (navneet1v) [inline server/src/main/java/org/opensearch/search/pipeline/Pipeline.java:121]: should we pass this context to SearchPhaseResults Processor too? I think we should. Whats your thought? @msfroh

File: server/src/main/java/org/opensearch/search/pipeline/SearchResponseProcessor.java
Comment (navneet1v) [inline server/src/main/java/org/opensearch/search/pipeline/SearchResponseProcessor.java:22]: can we have documentation on this function, same for the RequestProcessor too.

File: server/src/main/java/org/opensearch/search/pipeline/PipelinedRequest.java
Comment (navneet1v) [inline server/src/main/java/org/opensearch/search/pipeline/PipelinedRequest.java:26]: instead of having request context as Map<String,Object> should create a class and abstract this map inside the class? Which will make sure that pipelinedRequestContext is acting as container where if needed in future we can add more members without breaking the processor.process interface. 

Also this will ensure that if there are some context variable which we want the individual processors not to change we can have that.

File: modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/OversampleRequestProcessor.java
Comment (msfroh) [reply modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/OversampleRequestProcessor.java:47]: Good catch!

The default of 10 is set by `DefaultSearchContext`: https://github.com/opensearch-project/OpenSearch/blob/e5c4f9d2016bfa50c464945dc9533aa4aee49922/server/src/main/java/org/opensearch/search/DefaultSearchContext.java#L258

That executes after this. By default, this will get `originalSize = -1`.

I guess I can check for -1 and initialize it to 10 instead, though that's duplicating the logic from `DefaultSearchContext`. I can also move it out into a constant -- I'm a little surprised it's not already a constant.

Comment (msfroh) [reply modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/OversampleRequestProcessor.java:47]: Oh -- actually it is a constant, just not used by DefaultSearchContext:

https://github.com/opensearch-project/OpenSearch/blob/e98ded62d4b672a03acb4beb02bd047d7d4e1c69/server/src/main/java/org/opensearch/search/SearchService.java#L272

I think I'll change DefaultSearchContext to use that too.

File: modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/OversampleRequestProcessor.java
Comment (msfroh) [reply modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/OversampleRequestProcessor.java:49]: That max is an index setting and we don't (easily) have access to index settings in search processors. 

(Technically, on startup, the module gets a reference to `SearchPipelineService`, which exposes a reference to `ClusterService`, which exposes `ClusterState`, which has `Metadata`, including `IndexMetadata`.)

I think I'm okay with the processor doing what it was asked, and letting it throw an exception when it gets to `DefaultSearchContext.preProcess`. What do you think?

File: server/src/main/java/org/opensearch/search/pipeline/Pipeline.java
Comment (msfroh) [reply server/src/main/java/org/opensearch/search/pipeline/Pipeline.java:121]: Yes. I can make that change too.

File: server/src/main/java/org/opensearch/search/pipeline/PipelinedRequest.java
Comment (msfroh) [reply server/src/main/java/org/opensearch/search/pipeline/PipelinedRequest.java:26]: > instead of having request context as Map<String,Object> should create a class and abstract this map inside the class? Which will make sure that pipelinedRequestContext is acting as container where if needed in future we can add more members without breaking the processor.process interface later on.

Yeah... leaving it open to potential future changes without breaking existing code sounds like a good idea. 

> Also, it will allow individual processor to extend that PipelineContext class as per their processor needs too.

Not sure what you mean here. We'll need to define `PipelineContext` in core, in order to make sure everyone agrees on the thing we're passing around.

> Also this will ensure that if there are some context variable which we want the individual processors not to change we can have that.

Like values that we want processors to be able to read but not write? (I guess we could also have some "write-once" fields that throw an exception if another processor tries to change it.)

I'll wrap the map in a class (that will just hold the map for now).

File: modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/OversampleRequestProcessor.java
Comment (navneet1v) [reply modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/OversampleRequestProcessor.java:54]: I think thats fine. As long as some will throw the exception and don't let the request process.

Resolving the comment.

**Comment by harshavamsi** [comment]
@navneet1v @noCharger looks good?

**Comment by harshavamsi** [comment]
@reta can you also take a look?

File: server/src/main/java/org/opensearch/search/pipeline/StatefulSearchResponseProcessor.java
Comment (reta) [inline server/src/main/java/org/opensearch/search/pipeline/StatefulSearchResponseProcessor.java:25]: Should we have dedicated `PipelinedResponseContext responseContext` ? I think it would make sense to separate the request / response state passing.

File: server/src/main/java/org/opensearch/search/pipeline/PipelinedRequestContext.java
Comment (reta) [inline server/src/main/java/org/opensearch/search/pipeline/PipelinedRequestContext.java:20]: Should we hide the details that this is `Map`? May be have `put` / `get` / `getAll` APIs instead?

Comment (msfroh) [reply server/src/main/java/org/opensearch/search/pipeline/PipelinedRequestContext.java:20]: Maybe. 

I think I like the idea of having just `put` and `get`, since I would prefer that the API express the opinion that you should know what variables you're reading or writing.

Originally, the `Map` **was** the context object, but @navneet1v suggested wrapping it in a class for future extensibility in the form of added fields. For example, I'm considering the idea of optionally attaching the `IdentityService.getSubject()` to the context (resolved on the original REST action thread) so that it can be used by processors. (That would be a final, read-only context field.)

Anyway, I think I like the idea of encapsulating the map behind `void putGenericVariable(String, Object)` and `Object getGenericVariable(String)` methods. (Actually, since I was inspired by Java servlets, I could go with `putAttribute` and `getAttribute`.)

File: server/src/main/java/org/opensearch/search/pipeline/StatefulSearchResponseProcessor.java
Comment (msfroh) [reply server/src/main/java/org/opensearch/search/pipeline/StatefulSearchResponseProcessor.java:25]: The original motivation for this task was specifically to propagate state from request processors to response processors, like in the oversample/truncate processors in this PR. 

My original thought was to define a new kind of processor (a "bracket" processor) that would implement both the request and response processor interfaces (specifically to accomplish stuff like oversampling on request and undoing that oversampling on response). 

Of course, since processors had already been established to be stateless (and thread-safe for use across multiple requests), that design started to become really awkward. So, when @navneet1v suggested treating it as "passing state through the pipeline", I jumped on it.

So, I specifically want the same context to flow through the whole pipeline, from request processing through to response processing.

File: server/src/main/java/org/opensearch/search/pipeline/PipelinedRequestContext.java
Comment (msfroh) [reply server/src/main/java/org/opensearch/search/pipeline/PipelinedRequestContext.java]: I went with `getAttribute/setAttribute`. 

@reta, what do you think?

Comment (reta) [reply server/src/main/java/org/opensearch/search/pipeline/PipelinedRequestContext.java]: > @reta, what do you think?

Love it, thanks @msfroh !

File: server/src/main/java/org/opensearch/search/pipeline/StatefulSearchResponseProcessor.java
Comment (reta) [reply server/src/main/java/org/opensearch/search/pipeline/StatefulSearchResponseProcessor.java:25]: Got it, so basically there is nothing really request-specific here, but more like: here is the context, may be request processor set something or may be response processor did, we don't really care. May be we could go with  more generic name of `PipelinedProcessingContext` context` ?

Comment (msfroh) [reply server/src/main/java/org/opensearch/search/pipeline/StatefulSearchResponseProcessor.java:25]: Sure, that could work.

I'd like a name that makes it very clear that it's scoped to the lifetime of one request/response flow (i.e. it's not persistent).

**Comment by harshavamsi** [comment]
@reta looks good now?

**Comment by reta** [comment]
> @reta looks good now?

@harshavamsi https://github.com/opensearch-project/OpenSearch/pull/9405#issuecomment-1814980720

File: server/src/main/java/org/opensearch/search/pipeline/StatefulSearchResponseProcessor.java
Comment (msfroh) [reply server/src/main/java/org/opensearch/search/pipeline/StatefulSearchResponseProcessor.java]: I ended up going with `PipelineProcessingContext` 👍

File: modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/ScriptRequestProcessor.java
Comment (msfroh) [reply modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/ScriptRequestProcessor.java:104]: If I recall correctly, Painless script context is generally passed as a Map. In this case, we're adapting our custom class to the Map interface.

Comment (reta) [reply modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/ScriptRequestProcessor.java:104]: Oh I see, than we could just go with `AbstractMap`, right?

Comment (reta) [reply modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/ScriptRequestProcessor.java:104]: > I'm happy to make that change if we think it would be better than the `BasicMap` approach.

I see the picture now, thanks @msfroh, I thin we could keep it like that but reshuffle classes a bit:
 - move `BasicMap` to `org.opensearch.search.pipeline.common` and make it package private
 - move the `SearchRequestMap` to `org.opensearch.search.pipeline.common` and make it package private as well

What do you think?

**Review by reta — APPROVED** [review]

**Comment by austintlee** [comment]
Awesome!

**Comment by navneet1v** [comment]
This is awesome. an awaited feature. :D

**Pull request merged**
